akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlattenMergeSpec.scala

```diff
@@ -0,0 +1,129 @@
+        .flatMapMerge(4, identity)
+        .runWith(toSet)
+        .futureValue should ===((0 until 40).toSet)
+    }
+
+    "not be held back by one slow stream" in {
+      Source(List(src(0), src(1), blocked, src(2), src(3)))
+        .flatMapMerge(3, identity)
+        .take(40)
+        .runWith(toSet)
+        .futureValue should ===((0 until 40).toSet)
+    }
+
+    "respect breadth" in {
+      val seq = Source(List(src(0), src(1), src(2), blocked, blocked, src(3)))
+        .flatMapMerge(3, identity)
+        .take(40)
+        .runWith(toSeq)
+        .futureValue
+
+      seq.take(30).toSet should ===((0 until 30).toSet)
+      seq.drop(30).toSet should ===((30 until 40).toSet)
+    }
+
+    "propagate early failure from main stream" in {
+      val ex = new Exception("buh")
+      intercept[TestFailedException] {
+        Source.failed(ex)
+          .flatMapMerge(1, identity)
+          .runWith(Sink.head)
+          .futureValue
+      }.cause.get should ===(ex)
+    }
+
+    "propagate late failure from main stream" in {
+      val ex = new Exception("buh")
+      intercept[TestFailedException] {
+        (Source(List(blocked, blocked)) ++ Source.failed(ex))
+          .flatMapMerge(10, identity)
+          .runWith(Sink.head)
+          .futureValue
+      }.cause.get should ===(ex)
+    }
+
+    "bubble up substream exceptions" in {
+      val ex = new Exception("buh")
+      intercept[TestFailedException] {
+        Source(List(blocked, blocked, Source.failed(ex)))
+          .flatMapMerge(10, identity)
+          .runWith(Sink.head)
+          .futureValue
+      }.cause.get should ===(ex)
+    }
+
+    "cancel substreams when failing from main stream" in {
+      val p1, p2 = TestPublisher.probe[Int]()
+      val ex = new Exception("buh")
+      val p = Promise[Source[Int, Unit]]
+      (Source(List(Source(p1), Source(p2))) ++ Source(p.future))
+        .flatMapMerge(5, identity)
+        .runWith(Sink.head)
+      p1.expectRequest()
+      p2.expectRequest()
+      p.failure(ex)
+      p1.expectCancellation()
+      p2.expectCancellation()
+    }
+
+    "cancel substreams when failing from substream" in {
+      val p1, p2 = TestPublisher.probe[Int]()
+      val ex = new Exception("buh")
+      val p = Promise[Int]
+      Source(List(Source(p1), Source(p2), Source(p.future)))
+        .flatMapMerge(5, identity)
+        .runWith(Sink.head)
+      p1.expectRequest()
+      p2.expectRequest()
+      p.failure(ex)
+      p1.expectCancellation()
+      p2.expectCancellation()
+    }
+
+    "cancel substreams when being cancelled" in {
+      val p1, p2 = TestPublisher.probe[Int]()
+      val ex = new Exception("buh")
+      val sink = Source(List(Source(p1), Source(p2)))
+        .flatMapMerge(5, identity)
+        .runWith(TestSink.probe)
+      sink.request(1)
+      p1.expectRequest()
+      p2.expectRequest()
+      sink.cancel()
+      p1.expectCancellation()
+      p2.expectCancellation()
+    }
```

Comment:
maybe add a test to see make sure `flatMapMerge(5, throw Boom)` is handled correctly?